### ISSUE TITLE
Client components used only inside server actions are not added to the module graph

### ISSUE THREAD

**Issue by Dwlad90** [vercel/next.js#58125]
### Link to the code that reproduces this issue

https://github.com/Dwlad90/elegant-cloud-zckzqc/tree/draft/goofy-mountain

### To Reproduce

1. Install dependencies with `npm i`
2. Run the app in development mode (`next dev`)
3. Go to http://localhost:3000
4. See console error saying `" Error: Could not find the module "/Users/vladislavbuinovski/Projects/elegant-cloud-zckzqc/components/SA_CLNT_Component/Component.tsx#" in the React Client Manifest. This is probably a bug in the React Server Components bundler."`
5. As a result, some styles are missing. in `Server: Test server` (`Test server` should be pink with a yellow background)
6. Side-effect import of the client component or server action (as seen, commented out, on lines 5 and 6 in [ServerStyles.tsx](https://github.com/Dwlad90/elegant-cloud-zckzqc/blob/0f49a1193830b56534d782d93d513c55ff8ff741/components/ServerStyles/ServerStyles.tsx#L5-L6) force these modules into the graph, and causes the styles in 5 to be displayed

### Current vs. Expected behavior

Client side modules imported only in a server action module are not included in the React Client Manifest. They should be and the missing style should show up

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.1.0: Mon Oct  9 21:27:24 PDT 2023; root:xnu-10002.41.9~6/RELEASE_ARM64_T6000
Binaries:
  Node: 18.18.2
  npm: 10.1.0
  Yarn: 1.22.19
  pnpm: 8.9.2
Relevant Packages:
  next: 14.0.2-canary.18
  eslint-config-next: 14.0.2-canary.18
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
  ```

### Which area(s) are affected? (Select all that apply)

App Router

**Comment by roggc** [comment]
I am facing the same situation in an `Action` client component I defined, which what it does is wrap a server action in a `Suspense` component. This server action returns a client component. And then the error appears. The workaround is to include an import to the server action in a server component, `RootLayout` for example.

This is my `Action` client component:

```javascript
"use client";

import { Suspense, useState, useEffect } from "react";
import { usePropsChangedKey } from "@/app/hooks";

export default function Action({
  action,
  fallback = <>loading...</>,
  ...props
}) {
  const [JSX, setJSX] = useState(fallback);
  const propsChangedKey = usePropsChangedKey(...Object.values(props));

  useEffect(() => {
    setJSX(<Suspense fallback={fallback}>{action(props)}</Suspense>);
  }, [propsChangedKey, action]);

  return JSX;
}
```

This is the hook (for those who are curious about it):

```javascript
import { useState, useEffect } from "react";

export function usePropsChangedKey(...args) {
  const [propsChangedKey, setPropsChangedKey] = useState(0);

  useEffect(() => {
    setPropsChangedKey((k) => k + 1);
  }, [...args]);

  return propsChangedKey;
}
```

This is how you call the `Action` client component from a client or server component (client in this case):

```javascript
"use client";

import Action from "@/app/action";
import { greeting } from "@/app/actions/greeting";
import { useState } from "react";

export default function Client1() {
  const [userId, setUserId] = useState(1);

  return (
    <>
      <Action action={greeting} userId={userId} />
      <button
        onClick={() => {
          setUserId(2);
        }}
      >
        click
      </button>
    </>
  );
}
```

And this is the `greeting` server action code:

```javascript
"use server";

import Greeting from "@/app/action-components/greeting";
import MyError from "@/app/action-components/my-error";

const DELAY = 500;

const users = [
  { id: 1, username: "roggc" },
  { id: 2, username: "roger" },
];

export async function greeting({ userId }) {
  try {
    const username = await new Promise((r) => {
      setTimeout(() => {
        const user = users.find((u) => u.id === userId);
        if (user) {
          r(user.username);
        }
      }, DELAY);
    });

    // throw new Error("crash!");
    return <Greeting username={username} />;
  } catch (error) {
    return <MyError errorMessage={error.message} />;
  }
}
```

Now, without the workaround commented at the beginning, you get this error:

`⨯ Error: Could not find the module "C:\Users\roggc\dev\nextjs\test1\app\action-components\greeting.js#" in the React Client Manifest. This is probably a bug in the React Server Components bundler.
    at stringify (<anonymous>)`

The workaround, as I said, is to import the server action in a server component, for example the `RootLayout`:

```javascript
import { Inter } from "next/font/google";
import { greeting } from "@/app/actions/greeting"; // <-- this is necessary, if not fails to compile

const inter = Inter({ subsets: ["latin"] });

export const metadata = {
  title: "Create Next App",
  description: "Generated by create next app",
};

export default function RootLayout({ children }) {
  return (
    <html lang="en">
      <body className={inter.className}>{children}</body>
    </html>
  );
}
```

**Comment by JonathanVelkeneers** [comment]
I'm facing the same issue.
I am trying to use the AI-SDK-RSC to stream a client component in a server action.
This is the only place this client component is used so I'm also facing the "Could not find the module" error.
https://sdk.vercel.ai/docs/ai-sdk-rsc/streaming-user-interfaces

**Comment by obfuscatedgenerated** [comment]
Same for CSS modules for me

**Comment by nenadVes** [comment]
Yeah, client components are not in React Client Manifest when used in an server action that returns rendered JSX. Any updates on this issue? https://discord.com/channels/752553802359505017/1239912714210381865/1239913441464942682 posted here as well

**Comment by nenadVes** [comment]
Also, @roggc solution works, but puts all of the code into the client bundle which is not optimal and desired behaviour as we would like this portion of the code to be loaded on demand.

**Comment by jonathanhefner** [comment]
I recently ran into this issue.

Here is another reproduction using Next.js v15 and React v19: https://codesandbox.io/p/devbox/affectionate-tesla-w8ddmv.  When viewing the preview, notice that the "Get server component" button works, but the "Get client component" and "Get server component w/ client component" buttons trigger the aforementioned error.

<details>
  <summary>Environment information</summary>

  ```
  Operating System:
    Platform: linux
    Arch: x64
    Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
    Available memory (MB): 4102
    Available CPU cores: 2
  Binaries:
    Node: 20.9.0
    npm: 9.8.1
    Yarn: 1.22.19
    pnpm: 8.10.2
  Relevant Packages:
    next: 15.1.1-canary.23 // Latest available version is detected (15.1.1-canary.23).
    eslint-config-next: N/A
    react: 19.0.0
    react-dom: 19.0.0
    typescript: 5.3.3
  Next.js Config:
    output: N/A
  ```
</details>

**Comment by rikbrown** [comment]
This workaround no longer seems to work on 15.3.0 at least :(

**Comment by rikbrown** [comment]
Update: instead, I found you need to actually render a component from a file which then includes the rest of the tree. For example, instead of just `import "path/to/server/action.ts", in that file define:

```tsx
export function UseForceImportOfClientComponentTree() {
   return null
}
```

then in your layout, import this null component and render it. That's enough to then ensure all those client components are in the manifest.

**Comment by timneutkens** [comment]
Closing this issue in favor of #83186 which has a clear reproduction for the same issue.

**Comment by roggc** [comment]
For those interested in follow this pattern:

```javascript
"use client";

import {Suspense} from "react";
import {serverFunction} from "@/server/function/path";

export default function AComponent(){
  return <div>
    <Suspense>{serverFunction()}</Suspense>
  </div>;
}
```

where `serverFunction` it's a server function returning a Client Component, you can use it in [`Dinou`](https://dinou.dev).
